File: docker/hitl-ui/src/api/types/guardrails.ts
/**
 * TypeScript types for Guardrails Configuration System (P11-F01)
 *
 * These types match the backend Pydantic models for the guardrails
 * management API, including guidelines, evaluation, and audit logging.
 */

// ============================================================================
// Enum types
// ============================================================================

/**
 * Category of a guideline, matching backend GuidelineCategory enum.
 */
export type GuidelineCategory =
  | 'cognitive_isolation'
  | 'tdd_protocol'
  | 'hitl_gate'
  | 'tool_restriction'
  | 'path_restriction'
  | 'commit_policy'
  | 'custom';

/**
 * Type of action a guideline enforces.
 */
export type ActionType =
  | 'instruction'
  | 'tool_allow'
  | 'tool_deny'
  | 'hitl_require'
  | 'custom';

// ============================================================================
// Guideline condition and action
// ============================================================================

/**
 * Condition under which a guideline applies.
 * All fields are optional; a null or missing field means "match all".
 */
export interface GuidelineCondition {
  agents?: string[] | null;
  domains?: string[] | null;
  actions?: string[] | null;
  paths?: string[] | null;
  events?: string[] | null;
  gate_types?: string[] | null;
}

/**
 * Action to take when a guideline matches.
 */
export interface GuidelineAction {
  action_type: ActionType;
  instruction?: string | null;
  tools_allowed?: string[] | null;
  tools_denied?: string[] | null;
  gate_type?: string | null;
}

// ============================================================================
// Guideline entity
// ============================================================================

/**
 * A complete guideline as returned by the API.
 */
export interface Guideline {
  id: string;
  name: string;
  description: string;
  category: GuidelineCategory;
  priority: number;
  enabled: boolean;
  condition: GuidelineCondition;
  action: GuidelineAction;
  version: number;
  created_at: string;
  updated_at: string;
  created_by?: string | null;
  tenant_id?: string | null;
}

// ============================================================================
// API request/response types
// ============================================================================

/**
 * Paginated list response for guidelines.
 */
export interface GuidelinesListResponse {
  guidelines: Guideline[];
  total: number;
  page: number;
  page_size: number;
}

/**
 * Request body for creating a new guideline.
 */
export interface GuidelineCreateRequest {
  name: string;
  description?: string;
  category: GuidelineCategory;
  priority?: number;
  enabled?: boolean;
  condition: GuidelineCondition;
  action: GuidelineAction;
}

/**
 * Request body for updating an existing guideline.
 * The version field is required for optimistic locking.
 */
export interface GuidelineUpdateRequest {
  name?: string;
  description?: string;
  category?: GuidelineCategory;
  priority?: number;
  enabled?: boolean;
  condition?: GuidelineCondition;
  action?: GuidelineAction;
  version: number;
}

// ============================================================================
// Context evaluation types
// ============================================================================

/**
 * Request body for evaluating guidelines against a task context.
 */
export interface TaskContextRequest {
  agent?: string | null;
  domain?: string | null;
  action?: string | null;
  paths?: string[] | null;
  event?: string | null;
  gate_type?: string | null;
  session_id?: string | null;
}

/**
 * A single guideline that matched a task context evaluation.
 */
export interface EvaluatedGuideline {
  guideline_id: string;
  guideline_name: string;
  priority: number;
  match_score: number;
  matched_fields: string[];
}

/**
 * Response from evaluating guidelines against a task context.
 */
export interface EvaluatedContextResponse {
  matched_count: number;
  combined_instruction: string;
  tools_allowed: string[];
  tools_denied: string[];
  hitl_gates: string[];
  guidelines: EvaluatedGuideline[];
}

// ============================================================================
// Audit log types
// ============================================================================

/**
 * A single audit log entry.
 */
export interface AuditLogEntry {
  id: string;
  event_type: string;
  guideline_id?: string | null;
  timestamp: string;
  decision?: Record<string, unknown> | null;
  context?: Record<string, unknown> | null;
  changes?: Record<string, unknown> | null;
}

/**
 * Paginated response for audit log entries.
 */
export interface AuditLogResponse {
  entries: AuditLogEntry[];
  total: number;
}

// ============================================================================
// Filter/query parameter types
// ============================================================================

/**
 * Query parameters for listing guidelines.
 */
export interface GuidelinesListParams {
  category?: GuidelineCategory;
  enabled?: boolean;
  page?: number;
  page_size?: number;
}

/**
 * Query parameters for listing audit log entries.
 */
export interface AuditListParams {
  guideline_id?: string;
  event_type?: string;
  date_from?: string;
  date_to?: string;
  page?: number;
  page_size?: number;
}

// ============================================================================
// Import/export types
// ============================================================================

/**
 * Result of an import operation.
 */
export interface ImportResult {
  imported: number;
  errors: string[];
}

File: src/orchestrator/api/models/guardrails.py
"""Guardrails API data models.

Defines Pydantic models for guardrails CRUD operations,
evaluation, and audit log viewing.
"""

from __future__ import annotations

from datetime import datetime
from enum import Enum
from typing import Optional

from pydantic import BaseModel, Field


class GuidelineCategoryEnum(str, Enum):
    """Guideline category types.

    Values match the domain GuidelineCategory enum exactly to ensure
    lossless round-tripping between API and domain layers.
    """

    COGNITIVE_ISOLATION = "cognitive_isolation"
    HITL_GATE = "hitl_gate"
    TDD_PROTOCOL = "tdd_protocol"
    CONTEXT_CONSTRAINT = "context_constraint"
    AUDIT_TELEMETRY = "audit_telemetry"
    SECURITY = "security"
    CUSTOM = "custom"


class ActionTypeEnum(str, Enum):
    """Action type for guideline actions.

    Values match the domain ActionType enum exactly to ensure
    lossless round-tripping between API and domain layers.
    """

    INSTRUCTION = "instruction"
    TOOL_RESTRICTION = "tool_restriction"
    HITL_GATE = "hitl_gate"
    CONSTRAINT = "constraint"
    TELEMETRY = "telemetry"


class GuidelineConditionModel(BaseModel):
    """Condition for when a guideline applies."""

    agents: list[str] | None = None
    domains: list[str] | None = None
    actions: list[str] | None = None
    paths: list[str] | None = None
    events: list[str] | None = None
    gate_types: list[str] | None = None


class GuidelineActionModel(BaseModel):
    """Action to take when guideline matches."""

    action_type: ActionTypeEnum
    instruction: str | None = None
    tools_allowed: list[str] | None = None
    tools_denied: list[str] | None = None
    gate_type: str | None = None
    gate_threshold: str | None = None
    max_files: int | None = None
    require_tests: bool | None = None
    require_review: bool | None = None


class GuidelineCreate(BaseModel):
    """Request model for creating a new guideline."""

    name: str = Field(..., min_length=1, max_length=200)
    description: str = Field(default="", max_length=2000)
    category: GuidelineCategoryEnum
    priority: int = Field(default=100, ge=0, le=1000)
    enabled: bool = Field(default=True)
    condition: GuidelineConditionModel
    action: GuidelineActionModel


class GuidelineUpdate(BaseModel):
    """Request model for updating a guideline. All fields optional."""

    name: str | None = Field(default=None, min_length=1, max_length=200)
    description: str | None = Field(default=None, max_length=2000)
    category: GuidelineCategoryEnum | None = None
    priority: int | None = Field(default=None, ge=0, le=1000)
    enabled: bool | None = None
    condition: GuidelineConditionModel | None = None
    action: GuidelineActionModel | None = None
    version: int = Field(..., description="Version for optimistic locking")


class GuidelineResponse(BaseModel):
    """Response model for a single guideline."""

    id: str
    name: str
    description: str
    category: str
    priority: int
    enabled: bool
    condition: GuidelineConditionModel
    action: GuidelineActionModel
    version: int
    created_at: str
    updated_at: str
    created_by: str | None = None
    tenant_id: str | None = None


class GuidelinesListResponse(BaseModel):
    """Response model for listing guidelines."""

    guidelines: list[GuidelineResponse]
    total: int
    page: int = 1
    page_size: int = 20


class TaskContextRequest(BaseModel):
    """Request model for evaluating context."""

    agent: str | None = None
    domain: str | None = None
    action: str | None = None
    paths: list[str] | None = None
    event: str | None = None
    gate_type: str | None = None
    session_id: str | None = None


class EvaluatedGuidelineResponse(BaseModel):
    """Response for a single evaluated guideline."""

    guideline_id: str
    guideline_name: str
    priority: int
    match_score: float
    matched_fields: list[str]


class EvaluatedContextResponse(BaseModel):
    """Response model for evaluated context."""

    matched_count: int
    combined_instruction: str
    tools_allowed: list[str]
    tools_denied: list[str]
    hitl_gates: list[str]
    guidelines: list[EvaluatedGuidelineResponse]


class AuditLogEntry(BaseModel):
    """Response model for an audit log entry."""

    id: str
    event_type: str
    guideline_id: str | None = None
    timestamp: str
    decision: dict | None = None
    context: dict | None = None
    changes: dict | None = None


class AuditLogResponse(BaseModel):
    """Response model for listing audit entries."""

    entries: list[AuditLogEntry]
    total: int
